core/arch/arm/kernel/elf_load.c

```diff
@@ -619,24 +739,181 @@ TEE_Result elf_load_body(struct elf_load_state *state, vaddr_t vabase)
+	else
+		process_rel = e64_process_rel;
+
+	for (n = 0; n < ehdr.e_shnum; n++) {
+		uint32_t sh_type = get_shdr_type(state, n);
+
+		if (sh_type == SHT_REL || sh_type == SHT_RELA) {
+			res = process_rel(state, n, vabase);
+			if (res != TEE_SUCCESS)
+				return res;
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+/* Check the dynamic segment and save info for later */
+static TEE_Result e32_read_dynamic(struct elf_load_state *state,
+				   vaddr_t vabase)
+{
+	Elf32_Shdr *shdr = state->shdr;
+	struct elf_ehdr ehdr;
+	Elf32_Dyn *dyn_start;
+	Elf32_Dyn *dyn_end;
+	Elf32_Dyn *dyn;
+	vaddr_t dynstr = 0;
+	size_t dynstr_size = 0;
+	vaddr_t dynsym = 0;
+	size_t n;
+
+	copy_ehdr(&ehdr, state);
+
+	dyn_start = (Elf32_Dyn *)state->dyn;
+	if (!ALIGNMENT_IS_OK(dyn_start, Elf32_Dyn))
+		return TEE_ERROR_BAD_FORMAT;
+	dyn_end = dyn_start + state->dyn_size / sizeof(Elf32_Dyn);
+
+	/*
+	 * Find the address and size of the string table (.strtab)
+	 */
+	for (dyn = dyn_start; dyn < dyn_end; dyn++) {
+		if (dyn->d_tag == DT_STRTAB)
+			dynstr = dyn->d_un.d_ptr;
+		else if (dyn->d_tag == DT_STRSZ)
+			dynstr_size = dyn->d_un.d_val;
+	}
+	if (!dynstr || dynstr >= state->vasize)
+		return TEE_ERROR_BAD_FORMAT;
+	if (!dynstr_size || (dynstr + dynstr_size) > state->vasize)
+		return TEE_ERROR_BAD_FORMAT;
+	state->dynstr = (char *)(vabase + dynstr);
+	state->dynstr_size = dynstr_size;
+
+	/*
+	 * Find the .dynsym section (contains the global symbols).
+	 * There is an entry for it (DT_SYMTAB) in the .dynamic section, so we
+	 * could use the above loop to find the section address. Unfortunately,
+	 * the size information is not present. So, we have to parse the
+	 * section headers instead.
+	 */
+	for (n = 0; n < ehdr.e_shnum; n++)
+		if (shdr[n].sh_type == SHT_DYNSYM)
+			break;
+	if (n == ehdr.e_shnum)
+		return TEE_ERROR_BAD_FORMAT;
+	dynsym = shdr[n].sh_addr;
+	if (!dynsym || dynsym >= state->vasize)
+		return TEE_ERROR_BAD_FORMAT;
+	state->dynsym = (void *)(vabase + dynsym);
+	state->dynsym_size = shdr[n].sh_size;
+
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result e32_get_needed(struct elf_load_state *state,
+				 vaddr_t vabase, char ***names_ret,
+				 size_t *num_names_ret)
+
+{
+	Elf32_Dyn *dyn_start;
+	Elf32_Dyn *dyn_end;
+	Elf32_Dyn *dyn;
+	Elf32_Word offs;
+	TEE_Result res;
+	size_t num_names = 0;
+	char **names = NULL;
+
+	res = e32_read_dynamic(state, vabase);
+	if (res)
+		return res;
+
+	dyn_start = (Elf32_Dyn *)state->dyn;
+	dyn_end = dyn_start + state->dyn_size / sizeof(Elf32_Dyn);
+
+	/* Now look for needed libraries and fill output array */
+	for (dyn = dyn_start; dyn < dyn_end; dyn++) {
+		void *p;
+
+		if (dyn->d_tag != DT_NEEDED)
+			continue;
+		if (!state->dynstr) {
+			free(names);
+			return TEE_ERROR_BAD_FORMAT;
+		}
+		p = realloc(names, (num_names + 1) * sizeof(char *));
+		if (!p) {
+			free(names);
+			return TEE_ERROR_OUT_OF_MEMORY;
+		}
+		names = p;
+		offs = dyn->d_un.d_val;
+		names[num_names] = (char *)state->dynstr + offs;
```

Comment:
Do we need to check that `names[num_names]` is a valid pointer to a valid string?
Since this is a pointer into user space, should we guard against a lost terminating `\0` when parsing it?

Reply:
Or perhaps use a `copy_from_user_string()` function or similar. On a longer term we'll need to do the ELF loading from user space.